plugins/network-elements/nuage-vsp/src/com/cloud/network/manager/NuageVspManagerImpl.java

```diff
@@ -152,50 +243,179 @@ public NuageVspDeviceVO addNuageVspDevice(AddNuageVspDeviceCmd cmd) {
+                SyncNuageVspCmsIdAnswer answer = (SyncNuageVspCmsIdAnswer) _agentMgr.easySend(nuageVspDevice.getHostId(), syncCmd);
+                if (answer != null && answer.getSuccess()) {
+                    registerNewNuageVspDevice(cmsIdConfig, nuageVspDevice.getId() + ":" + answer.getNuageVspCmsId());
+
+                    detail = new DetailVO(host.getId(), "nuagevspcmsid", answer.getNuageVspCmsId());
+                    _hostDetailsDao.persist(detail);
+                    hostdetails.put("nuagevspcmsid", answer.getNuageVspCmsId());
+                    resource.configure((String) hostdetails.get("hostname"), hostdetails);
+
+                    auditDomainsOnVsp((HostVO) host, true, false);
+                } else {
+                    throw new CloudRuntimeException("Failed to register CMS ID");
+                }
+                return nuageVspDevice;
+            } else {
+                throw new CloudRuntimeException("Failed to add Nuage Vsp Device due to internal error.");
+            }
+        } catch (Exception e) {
+            throw new CloudRuntimeException(e);
+        }
+    }
+
+    @Override
+    public NuageVspDeviceVO updateNuageVspDevice(UpdateNuageVspDeviceCmd command) {
+        ServerResource resource = new NuageVspResource();
+        final String deviceName = Network.Provider.NuageVsp.getName();
+        ExternalNetworkDeviceManager.NetworkDevice networkDevice = ExternalNetworkDeviceManager.NetworkDevice.getNetworkDevice(deviceName);
+        final Long physicalNetworkId = command.getPhysicalNetworkId();
+        PhysicalNetworkVO physicalNetwork = _physicalNetworkDao.findById(physicalNetworkId);
+        if (physicalNetwork == null) {
+            throw new InvalidParameterValueException("Could not find phyical network with ID: " + physicalNetworkId);
+        }
+
+        final PhysicalNetworkServiceProviderVO ntwkSvcProvider = _physicalNetworkServiceProviderDao.findByServiceProvider(physicalNetwork.getId(),
+                networkDevice.getNetworkServiceProvder());
+        if (ntwkSvcProvider == null) {
+            throw new CloudRuntimeException("Network Service Provider: " + networkDevice.getNetworkServiceProvder() + " is not enabled in the physical network: "
+                    + physicalNetworkId + "to add this device");
+        } else if (ntwkSvcProvider.getState() == PhysicalNetworkServiceProvider.State.Shutdown) {
+            throw new CloudRuntimeException("Network Service Provider: " + ntwkSvcProvider.getProviderName() + " is in shutdown state in the physical network: "
+                    + physicalNetworkId + "to add this device");
+        }
+
+        HostVO nuageVspHost = null;
+        NuageVspDeviceVO nuageVspDevice = null;
+        List<NuageVspDeviceVO> nuageVspDevices = _nuageVspDao.listByPhysicalNetwork(physicalNetworkId);
+        if (nuageVspDevices == null || nuageVspDevices.size() == 0) {
```

Comment:
Please ``List#isEmpty()`` rather a size check as it is the idiomatic way to check that a list is empty and more clearly expresses intent.